### PR TITLE
Rockchip: add H96 Max+ support

### PR DESCRIPTION
**Pull request by easybe**
The device tree is based on DTS files for similar devices supported by
LibreELEC and the DTB extracted from the H96 Max+ Android 9 image
(RK3328_DC_H96_Max+_HS2734C_9_20191031.2215.img).

Unfortunately, devfreq is not available for the GPU (lima dirver) yet.
As a workaround, we simply do not configure pwm1 and the corresponding
regulator.

### CODE REVIEW

**Comment by easybe** [comment]
This patch can probably still be improved. I'd be happy to get some feedback.

However, following command produces an image that works for me and my H96 Max+ TV box:
```
PROJECT=Rockchip DEVICE=RK3328 ARCH=arm UBOOT_SYSTEM=box-h96mp make image
```
I am still trying to figure out why HDMI-CEC is not working. I also ordered the similar T9 box to rule out a hardware issue (and add support for that device).

**Comment by Kwiboo** [comment]
@easybe Thanks for your contribution!

I am working on a rockchip 5.5 patchset, should be done any day now, can I include your linux devicetree commit as part of the general rockchip 5.5 patchset update?
If not please wait for the 5.5 update PR to land and rebase your PR on top of that.

**Comment by easybe** [comment]
@Kwiboo go ahead, that would be awesome, thanks!

Waiting and rebasing would be fine for me too though.

**Comment by easybe** [comment]
@Kwiboo just make sure you have the latest version of my patch when you integrate it. I can not guarantee that this is the final one 🙊 FYI, following branch should always be up to date: https://github.com/easybe/linux/tree/h96mp

**Comment by easybe** [comment]
Even though Linux seems to work just fine with this DT, it turns out that the box does not have an rk805 (or similar) PMIC. I tried to adapt the regulator part according to the Android DT (https://github.com/easybe/linux/commit/dc7cafab873f28c025b2843109766ae057c7fbc9). However, Kodi does not start now and I get a kernel panic:
```
[   79.837738] WARNING: CPU: 3 PID: 0 at kernel/workqueue.c:1469 __queue_work+0x424/0x4e0
[   79.837746] Modules linked in: hantro_vpu(C) videobuf2_vmalloc videobuf2_dma_contig videobuf2_memops ir_nec_decoder rc_x96max v4l2_mem2mem gpio_ir_recv videobuf2_v4l2 videobuf2_common videodev mali(O) mc wireguard(O) ip6_udp_tunnel udp_tunnel ipv6 nf_defrag_ipv6
[   79.837875] CPU: 3 PID: 0 Comm: swapper/3 Tainted: G         C O      5.5.0 #1
[   79.837881] Hardware name: Rockchip RK3328 TV box H96 Max+ (DT)
[   79.837896] pstate: 20000085 (nzCv daIf -PAN -UAO)
[   79.837910] pc : __queue_work+0x424/0x4e0
[   79.837920] lr : __queue_work+0x170/0x4e0
[   79.837928] sp : ffff80001001bc80
[   79.837934] x29: ffff80001001bc80 x28: ffff0000ed4e1c80 
[   79.837948] x27: 0000000000000003 x26: ffff0000f0409a00 
[   79.837962] x25: ffff8000111d1018 x24: 000000000000002a 
[   79.837976] x23: ffff800011359528 x22: 0000000000000004 
[   79.837989] x21: ffff0000fdc24000 x20: ffff0000ed4e1c98 
[   79.838003] x19: ffff0000eb4a6b00 x18: 0000000000000000 
[   79.838016] x17: 0000000000000000 x16: 0000000000000000 
[   79.838026] x15: 0000000000000000 x14: 0000000016c66563 
[   79.838039] x13: 0032dcd555555555 x12: 000032dcd555bd80 
[   79.838051] x11: ffff0000fe7b18f0 x10: 0000000000000040 
[   79.838065] x9 : ffff800011213b48 x8 : ffff800011213b40 
[   79.838078] x7 : 0000000000000000 x6 : 0000000000000001 
[   79.838089] x5 : 000000000000003f x4 : 0000000000000001 
[   79.838100] x3 : 0000000000000000 x2 : 0000000000000002 
[   79.838113] x1 : ffff00003fc00003 x0 : ffff0000ed4e1ca0 
[   79.838126] Call trace:
[   79.838138]  __queue_work+0x424/0x4e0
[   79.838149]  queue_work_on+0xd8/0xf0
[   79.838316]  _mali_osk_wq_schedule_work+0x24/0x30 [mali]
[   79.838447]  mali_group_timeout+0x38/0x68 [mali]
[   79.838467]  call_timer_fn+0x34/0x170
[   79.838483]  __run_timers.part.0+0x200/0x330
[   79.838496]  run_timer_softirq+0x40/0x78
[   79.838514]  __do_softirq+0x168/0x38c
[   79.838528]  irq_exit+0xd0/0x100
[   79.838547]  __handle_domain_irq+0x6c/0xc0
[   79.838558]  gic_handle_irq+0x58/0xa8
[   79.838571]  el1_irq+0xbc/0x140
[   79.838588]  arch_cpu_idle+0x40/0x1c8
[   79.838605]  default_idle_call+0x24/0x44
[   79.838620]  do_idle+0x214/0x298
[   79.838631]  cpu_startup_entry+0x2c/0x30
[   79.838650]  secondary_start_kernel+0x124/0x160
``` 

Any ideas what might go wrong now?

[dmesg.txt](https://github.com/LibreELEC/LibreELEC.tv/files/4274310/dmesg.txt)

**Comment by easybe** [comment]
OK, it looks like the PWM regulators are not working as expected. I will try to fix it.

**Comment by easybe** [comment]
As far a I can tell, it seems that vdd_logic is too low. I measured ~925mV and and in sysfs `microvolts` is set to 900000 . When running Android I measure something closer to 1.05V. I assume that the GPU should set the voltage. @Kwiboo can you maybe tell me how GPU frequency scaling is enabled/disabled? Or why it might not be enabled in my case? I do not see Linux setting the voltage at boot or later. We could also make U-Boot set the correct voltage, it would at least make the box fully boot, I guess.

**Comment by easybe** [comment]
So, it turns out that we currently do not have devfreq for the lima driver. I added a workaround for the time being. As far as I can tell, the current DT describes the hardware pretty well and IMHO is good enough for testing.

**Comment by Ntemis** [comment]
@easybe please remove the pine64 copyright notice in the device tree

**Comment by fransvanberckel** [comment]
I have a H96 Max, without a plus. I extracted my DTS file by the next steps ...

Install on Android  the Termux app.
Next click on the icon starting the app.
Become root by # su -
Change directory by # cd /data
Make a directory by # mkdir root
Change directory by # cd root
You can save some data here.
Make a directory by # mkdir device-tree-copy
Change directory by # cd device-tree-copy
And run the next command ...
# for i in `find /proc/device-tree/ -type d | sed 's,/proc/device-tree/,,g'` ; do echo $i; mkdir -p $i ; done && for i in `find /proc/device-tree/ -type f | sed 's,/proc/device-tree/,,g'` ; do echo $i; cat /proc/device-tree/$i > $i ; done
Change directory by # cd ..
Pack your directory by # tar czf device-tree-copy.tar.gz device-tree-copy
Next copy the device tree tar file to your usb stick
At your Linux computer you can build a dts file, by extracting the tar file and running ...
# dtc -I fs -O dts device-tree-copy

**Comment by easybe** [comment]
@fransvanberckel I am not sure what you are trying to tell me. I am well aware how to extract a DTB.

Anyway, the job is already done. This is a working patch. It is not simply a converted DT, but, handcrafted to work with mainline Linux.

The H96 Max uses an RK3318 SoC. AFAIK RK3318 is not yet supported by mainline Linux. So, that will take a while...

**Comment by easybe** [comment]
Regarding devfreq for lima. We could try out https://patchwork.kernel.org/patch/11448015/. Or, we can just wait. At the moment I am working on other things.

**Comment by fransvanberckel** [comment]
> So, it turns out that we currently do not have devfreq for the lima driver. I added a workaround for the time being. As far as I can tell, the current DT describes the hardware pretty well and IMHO is good enough for testing.

Keep an eye on - devfreq fixes for panfrost, - drm: lima: devfreq and cooling device support ...
https://patchwork.kernel.org/project/linux-amlogic/list/?submitter=38211

**Comment by fransvanberckel** [comment]
> @fransvanberckel the job is already done. This is a working patch. It is not simply a decompiled DT, but, handcrafted to work with mainline Linux.
> 
> The H96 Max uses an RK3318 SoC. AFAIK RK3318 is not yet supported by mainline Linux. So, I guess that will take a while...

About H96 Max uses and RK3318 SoC. If that's true, why does the dts file call ... compatible = "rockchip,rk3328-box-avb", "rockchip,rk3328", "rockchip,rk3318", "rockchip,rk3318-evb" ;-)

**Comment by easybe** [comment]
Then I guess the RK3328 and RK3318 are very similar. Maybe it will just work. AFAIK the RK3318 is  the low-cost version of the RK3328. What that exactly means, I don't know.

**Comment by fransvanberckel** [comment]
With Android, # inxi about the machine does, system: Rockchip RK3328 box

**Comment by easybe** [comment]
Do you have this one https://www.aliexpress.com/item/33037040559.html ?

I am actually working on a display driver for boxes like that one.

**Comment by fransvanberckel** [comment]
Yes i have.

File: projects/Rockchip/patches/linux/default/linux-0006-for-libreelec-h96mp.patch [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+ create mode 100644 arch/arm64/boot/dts/rockchip/rk3328-box-h96mp.dts
+
+diff --git a/arch/arm64/boot/dts/rockchip/Makefile b/arch/arm64/boot/dts/rockchip/Makefile
+index c475e55a888c..3dac42051486 100644
+--- a/arch/arm64/boot/dts/rockchip/Makefile
++++ b/arch/arm64/boot/dts/rockchip/Makefile
+@@ -4,6 +4,7 @@ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3308-evb.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3308-roc-cc.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3328-a1.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3328-box.dtb
```
Comment (easybe) [inline projects/Rockchip/patches/linux/default/linux-0006-for-libreelec-h96mp.patch:28]: @Kwiboo I noticed that these *.dtb files were removed with bc55a55cbbb91e3fbc38cd52e73436def957772d. Is there a special reason for this? Do you need help porting them?

**Comment by CvH** [comment]
closing this,
if this is still needed pls update the pr and we can open the pr again

**Comment by aqzina** [comment]
does this patch not being merged mean h96 max+ devices will not be supported on the next official release yet? Or is there another dtb file that will be usable instead?

**Comment by easybe** [comment]
I am working on an update of this patch and see if we can integrate it. Worst case you will have to add the DTB file manually.

**Comment by easybe** [comment]
@CvH I updated my branch. The changes still need to be tested it though. Will do that as soon as I have access to the hardware again.

**Comment by easybe** [comment]
OK, tested. AFAICT everything seems to work fine. @CvH can this PR be reopened?

**Comment by devidasjadhav** [comment]
@easybe want to help test this but I was unable to boot with this patch.

I have built on 10.0 with your patch.

there was one error for
```diff
+&usbdrd_dwc3 {
+	dr_mode = "host";
+	status = "okay";
+};
```
used [lkml_patch](https://patchwork.kernel.org/project/linux-rockchip/patch/20171204094041.25439-2-heiko@sntech.de/)

with this combination I was able to build.

but I was unable to boot.

    PROJECT=Rockchip DEVICE=RK3328 ARCH=arm UBOOT_SYSTEM=h96mp make image

I have [H96MAX + 3318](https://www.h96tvbox.com/Android-box-Brand-H96-Max-Tv-Box/96-656-ram4gb-and-rom-32gb-android-10-tv-box.html).

I don't have serial port soldered yet I will be trying that and will report.

Method used to test.
flashed 16gb sd card and 8gb usb with build generated.
plugged in sd card and usb
powered on.


NOTE: This is my first try to boot anything on this box.

**Comment by easybe** [comment]
> but I was unable to boot.

I don't know what the current status is, but AFAIK the RK3318 is not supported:

https://forum.libreelec.tv/thread/23857-rockchip-dtb-file-for-hk1-max-cpu-rk3318-ram4g/?postID=154792#post154792

**Comment by devidasjadhav** [comment]
Thanks for the lead.
I tried to boot from sd card.
Until sd card was not booting. with red light. indicating boot error.

but I have changed trust.img, MiniLoaderAll.bin (idbloader.img), uboot.img.

extracted from [android image published](https://www.h96tvbox.com/content/6-Firmware-upgrade).
using procedure from [T firefly fw unpack documentation](https://wiki.t-firefly.com/ROC-RK3328-CC/fw_pack_unpack.html)

now Its booting from sd card but booting emmc.

My analysis from [boot wiki](http://opensource.rock-chips.com/wiki_Boot_option)

idbloader is not proper. because thats last change before successful boot.

@easybe  where did idbloader is created? where I can change emmc -> sd card.

**Pull request closed without merging**